packages/legend-query-builder/src/components/QueryLoader.tsx

```diff
@@ -0,0 +1,596 @@
+                              disabled={
+                                query.owner !== queryLoaderState.currentUserId
+                              }
+                            >
+                              <MenuContentItemLabel>
+                                Rename
+                              </MenuContentItemLabel>
+                            </MenuContentItem>
+                          )}
+                        </MenuContent>
+                      }
+                      menuProps={{
+                        anchorOrigin: {
+                          vertical: 'bottom',
+                          horizontal: 'left',
+                        },
+                        transformOrigin: {
+                          vertical: 'top',
+                          horizontal: 'left',
+                        },
+                        elevation: 7,
+                      }}
+                    >
+                      <MoreVerticalIcon />
+                    </DropdownMenu>
+                    <ChevronRightIcon />
+                  </div>
+                </div>
+              ))}
+              {queryLoaderState.queries.length === 0 && (
+                <BlankPanelContent>
+                  Search by name to find the relevant queries
+                </BlankPanelContent>
+              )}
+              {queryLoaderState.queries.length >=
+                DEFAULT_TYPEAHEAD_SEARCH_LIMIT && (
+                <div className="query-loader__result__section__message">
+                  Too many matching queries found, truncating list to be more
+                  specific
+                </div>
+              )}
+            </>
+          )}
+          <PanelLoadingIndicator
+            isLoading={
+              !queryLoaderState.loadQueriesState.hasCompleted ||
+              renameQueryState.isInProgress
+            }
+          />
+          {!queryLoaderState.loadQueriesState.hasCompleted && (
+            <BlankPanelContent>Loading queries...</BlankPanelContent>
+          )}
+        </div>
+      </>
+    );
+  },
+);
+
+export const QueryLoader = observer(
+  (props: {
+    queryLoaderState: QueryLoaderState;
+    graphManager: AbstractPureGraphManager;
+    loadQuery: (selectedQuery: LightQuery) => void;
+    renameQuery?:
+      | ((
+          selectedQuery: LightQuery,
+          updatedQueryName: string,
+          renameQueryState: ActionState,
+        ) => Promise<void>)
+      | undefined;
+    modalTitle?: string | undefined;
+    moreOptions?:
+      | {
+          loadAsDialog?: boolean;
+          isDeleteSupported?: boolean;
+          includeDefaultQueries?: boolean;
+        }
+      | undefined;
+    currentQueryId?: string | undefined;
+  }) => {
+    const {
+      queryLoaderState,
+      graphManager,
+      loadQuery,
+      renameQuery,
+      modalTitle,
+      moreOptions,
+      currentQueryId,
+    } = props;
+    if (moreOptions?.loadAsDialog) {
```

Comment:
I didn't like this pattern at first, but the more I think about it, the more sense i makes, especially with things like modal/popover, etc.